src/code/models/simulation.coffee

```diff
@@ -89,10 +97,49 @@ module.exports = class Simulation
 
     @reportFunc(data)
 
+  # Tests that the graph contains no loops consisting of dependent variables.
+  # A graph such as A->B<->C is invalid if B and C connect to each other and
+  # neither are accumulators
+  graphIsValid: ->
+    nodesSeenInSegment = []
+    _.each @nodes, (node) -> node._isValid = null
+
+    # Recursive function. We go throw a node's non-independent ancestors, and if
+    # we find ourselves again, or if any of our ancestors find themselves again,
+    # we have a loop.
+    isInALoop = (node) ->
+      linksIn = node.inLinks()
+      # quick exit if we're not a dependent variable, or if we've already been checked
+      return false if node._isValid or node.isAccumulator or linksIn.length is 0
+      for seen in nodesSeenInSegment
+        if seen is node
+          return true
+
+      nodesSeenInSegment.push node
+
+      for link in linksIn
+        return true if isInALoop link.sourceNode
+
+      return false
+
+    for node in @nodes
+      nodesSeenInSegment = []
```

Comment:
Does this declaration mask the array of the same-name on line 104? I feel like this only effects marking the individual nodes as invalid -- so its probably not critical.